File: android/src/main/java/expo/modules/tiktokopensdk/ExpoTikTokOpenSDKModule.kt
package expo.modules.tiktokopensdk

import android.content.Context
import com.bytedance.sdk.open.tiktok.TikTokOpenSDK
import com.bytedance.sdk.open.tiktok.api.TikTokOpenApi
import com.bytedance.sdk.open.tiktok.share.Share
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

class ExpoTikTokOpenSDKModule : Module() {
    private val context: Context
        get() = requireNotNull(appContext.reactContext)

    override fun definition() = ModuleDefinition {
        Name("ExpoTikTokOpenSDK")

        Function("share") { mediaUrls: List<String>, isImage: Boolean, isGreenScreen: Boolean, hashtags: List<String>, description: String ->
            return@Function try {
                val request = Share.Request()
                
                // Convert URLs to local files
                val mediaFiles = mediaUrls.map { File(it) }
                
                // Configure media type and files
                if (isImage) {
                    request.setImageList(mediaFiles)
                } else {
                    request.setVideoList(mediaFiles)
                }
                
                // Set hashtags and description
                if (hashtags.isNotEmpty()) {
                    request.setHashtagList(hashtags)
                }
                if (description.isNotEmpty()) {
                    request.setTitle(description)
                }
                
                // Configure green screen if needed
                if (isGreenScreen) {
                    request.setShareFormat(Share.SHARE_FORMAT_GREEN_SCREEN)
                }
                
                // Send share request
                val api = TikTokOpenApi.getInstance()
                val response = api.share(context, request)
                
                if (response.isSuccess()) {
                    mapOf(
                        "isSuccess" to true
                    )
                } else {
                    mapOf(
                        "isSuccess" to false,
                        "errorCode" to response.errorCode,
                        "subErrorCode" to response.subErrorCode,
                        "errorMsg" to (response.errorMsg ?: "Unknown error")
                    )
                }
            } catch (e: Exception) {
                mapOf(
                    "isSuccess" to false,
                    "errorCode" to -1,
                    "errorMsg" to (e.message ?: "Unknown error")
                )
            }
        }

        Function("auth") { permissions: List<String>?, state: String? ->
            return@Function try {
                val request = com.bytedance.sdk.open.tiktok.authorize.AuthRequest()
                permissions?.let { request.scope = it.joinToString(",") }
                state?.let { request.state = it }
                
                val api = TikTokOpenApi.getInstance()
                val response = api.authorize(context, request)
                
                if (response.isSuccess()) {
                    mapOf(
                        "isSuccess" to true,
                        "authCode" to (response.authCode ?: ""),
                        "state" to (response.state ?: ""),
                        "grantedPermissions" to (response.grantedPermissions ?: emptyList<String>())
                    )
                } else {
                    mapOf(
                        "isSuccess" to false,
                        "errorCode" to response.errorCode,
                        "errorMsg" to (response.errorMsg ?: "Unknown error")
                    )
                }
            } catch (e: Exception) {
                mapOf(
                    "isSuccess" to false,
                    "errorCode" to -1,
                    "errorMsg" to (e.message ?: "Unknown error")
                )
            }
        }

        OnCreate {
            // Initialize TikTok SDK when the module is created
            val appContext = requireNotNull(appContext.reactContext)
            TikTokOpenSDK.init(appContext)
        }
    }
}

File: ios/ExpoTikTokOpenSDKModule.swift
import ExpoModulesCore
import TikTokOpenSDK

public class ExpoTikTokOpenSDKModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoTikTokOpenSDK")

        Function("share") { (mediaUrls: [String], isImage: Bool, isGreenScreen: Bool, hashtags: [String], description: String) -> [String: Any] in
            return try await withPromise { promise in
                // Convert URLs to local file URLs if needed
                let urls = mediaUrls.map { URL(fileURLWithPath: $0) }
                
                // Create share request
                let request = TikTokOpenSDKShareRequest()
                
                // Configure media type and URLs
                if isImage {
                    request.mediaType = .image
                    request.localIdentifiers = urls.map { $0.path }
                } else {
                    request.mediaType = .video
                    request.localIdentifiers = urls.map { $0.path }
                }
                
                // Set hashtags and description
                if !hashtags.isEmpty {
                    request.hashtags = hashtags
                }
                if !description.isEmpty {
                    request.landedPageTitle = description
                }
                
                // Configure green screen if needed
                if isGreenScreen {
                    request.shareFormat = .greenScreen
                }
                
                // Send share request
                request.send { response in
                    if response.isSucceed {
                        promise.resolve([
                            "isSuccess": true
                        ])
                    } else {
                        promise.resolve([
                            "isSuccess": false,
                            "errorCode": response.errorCode,
                            "shareState": response.shareState?.rawValue ?? 0,
                            "errorMsg": response.errorMsg ?? "Unknown error"
                        ])
                    }
                }
            }
        }

        Function("auth") { (permissions: [String]?, state: String?) -> [String: Any] in
            return try await withPromise { promise in
                let request = TikTokOpenSDKAuthRequest()
                if let perms = permissions {
                    request.permissions = perms
                }
                if let st = state {
                    request.state = st
                }
                
                request.send { response in
                    if response.isSucceed {
                        promise.resolve([
                            "isSuccess": true,
                            "authCode": response.code ?? "",
                            "state": response.state ?? "",
                            "grantedPermissions": response.grantedPermissions ?? []
                        ])
                    } else {
                        promise.resolve([
                            "isSuccess": false,
                            "errorCode": response.errorCode,
                            "errorMsg": response.errorMsg ?? "Unknown error"
                        ])
                    }
                }
            }
        }

        // Lifecycle events
        OnCreate {
            // Initialize TikTok SDK when the module is created
            if let clientKey = self.appContext?.config?["iosClientKey"] as? String {
                TikTokOpenSDKApplicationDelegate.sharedInstance().registerApp(universalLink: nil, clientKey: clientKey)
            }
        }
    }
}
